### PR TITLE
Adding notification publisher strategies

### PR DESCRIPTION
**Pull request by jbogard**
I did not pull *all* the possible options from @remcoros 's samples. Just the basic couple options.

### CODE REVIEW

**Comment by jbogard** [comment]
This PR introduces the idea of notification publishers as a separate class to be injected into the `Mediator` class. Instead of only being able to override the `PublishCore` method in a derived class, this lets you inject your own implementation.:

```diff
public class Mediator : IMediator
{
    public Mediator(IServiceProvider serviceProvider) 
-        => _serviceProvider = serviceProvider;
+        : this(serviceProvider, new ForeachAwaitPublisher()) { }

+    public Mediator(IServiceProvider serviceProvider, INotificationPublisher publisher)
+    {
+        _serviceProvider = serviceProvider;
+        _publisher = publisher;
+    }
```

Two new publishing strategies added:
- `ForeachAwaitPublisher` (the default, existing strategy)
- `TaskWhenAllPublisher` - uses `Task.WhenAll` to publish

The publishing strategies now can accept the handler instance, for cases when ordering is needed:

```diff
+public interface INotificationPublisher
+{
+    Task Publish(IEnumerable<NotificationHandlerExecutor> handlerExecutors, INotification notification,
+        CancellationToken cancellationToken);
+ }
+ public record NotificationHandlerExecutor(object HandlerInstance, Func<INotification, CancellationToken, Task> HandlerCallback);
```

This required a breaking change in the `PublishCore` method in `Mediator`:

```diff
public class Mediator : IMediator {
-    protected virtual async Task PublishCore(IEnumerable<Func<INotification, CancellationToken, Task>> allHandlers, INotification notification, CancellationToken cancellationToken)
-    {
-        foreach (var handler in allHandlers)
-        {
-            await handler(notification, cancellationToken).ConfigureAwait(false);
-        }
-    }
+    protected virtual Task PublishCore(IEnumerable<NotificationHandlerExecutor> handlerExecutors, INotification notification, CancellationToken cancellationToken) 
+       => _publisher.Publish(handlerExecutors, notification, cancellationToken);
}
```

The notification publisher can be configured via `MediatrServiceConfiguation`:

```diff
+    public INotificationPublisher NotificationPublisher { get; set; } = new ForeachAwaitPublisher();
+    public Type? NotificationPublisherType { get; set; }
```

The `NotificationPublisherType` value overrides the instance property.

File: src/MediatR/NotificationPublishers/TaskWhenAllPublisher.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+/// return Task.WhenAll(tasks);
+/// </code>
+/// </summary>
+public class TaskWhenAllPublisher : INotificationPublisher
+{
+    public Task Publish(IEnumerable<NotificationHandlerExecutor> handlerExecutors, INotification notification, CancellationToken cancellationToken)
+    {
+        var tasks = handlerExecutors
+            .Select(handler => handler.HandlerCallback(notification, cancellationToken))
+            .ToList();
```
Comment (remcoros) [inline src/MediatR/NotificationPublishers/TaskWhenAllPublisher.cs:24]: tiny nit: using '.ToArray()' hits the most optimized path of Task.WhenAll. 

(passing an IEnumerable/ICollection to Task.WhenAll internally creates another array of tasks, discarding the given list, passing a Task array directly or using the params overload does not, saving an allocation)

**Comment by remcoros** [comment]
awesome!

**Comment by neozhu** [comment]
Will this feature be included in the next version? Very much looking forward to

**Pull request merged**